### DISCUSSION TITLE
How does Next.js render pages when in dev and prod?

### DISCUSSION THREAD

**Discussion by zcarc** [vercel/next.js#40650] [Help] [answered]
I wonder how Next.js renders when developing in "dev mode".

1. Run the app with "npm run dev" in dev mode
If the page doesn't have getStaticProps(SSG) or getServerSideProps(SSR) or ISG, by default how does rendering proceed in Next.js? In the official documentation, it is written that SSG is used by default and is enabled during build, but I am curious how it works in dev mode.

2. By default, "next/link" is rendered in CSR (https://nextjs.org/docs/api-reference/next/link). 
How is it rendered with or without SSG, SSR, and ISG on the page?

3. If I go to the page directly with the URL in the browser address bar without moving the page to "next/link", how do you render it with or without SSG, SSR, and ISG?

4. The above questions are when you do it in dev mode, but I wonder if there is a difference in rendering with the above questions when using prod.

**Comment by icyJoseph** [comment, marked as answer]
> 1. Run the app with "npm run dev" in dev mode
>    If the page doesn't have getStaticProps(SSG) or getServerSideProps(SSR) or ISG, by default how does rendering proceed in Next.js? In the official documentation, it is written that SSG is used by default and is enabled during build, but I am curious how it works in dev mode.

- https://nextjs.org/docs/basic-features/data-fetching/get-static-props#runs-on-every-request-in-development
> In development (next dev), getStaticProps will be called on every request.

Regarding pages that are optimised, without `getStaticProps` or `getServerSideProps`, these are also rendered on demand, as you navigate to them.

> 2. By default, "next/link" is rendered in CSR (https://nextjs.org/docs/api-reference/next/link).
>    How is it rendered with or without SSG, SSR, and ISG on the page?

The `next/link` component is an enabler of SPA transitions. Meaning that as far as the HTML is concerned, as long as it renders an anchor tag, then it is all good. 

The trick is that during CSR, when the user clicks on a link, instead of unloading the page, the client side router takes care of un-mounting the React Component Page, and mounting which ever React Component Page applies. 

So `next/link` should be implemented so that it renders an anchor tag. 

Links have an additional property which is `prefetch`, they prefetch JS and initial props (as JSON), for pages with `getStaticProps`, and only JS for pages with `getServerSideProps`. 

If you SPA transition into a page without data-fetching functions, the JS for the page is downloaded, and it is client side rendered. Though the presence of middleware has funny implications for this, it is still roughly the same.

> 3. If I go to the page directly with the URL in the browser address bar without moving the page to "next/link", how do you render it with or without SSG, SSR, and ISG?

If the `page doesn't exist`, it is rendered on the server, and the HTML is sent down to the client. If there's data fetching functions, a JSON payload is included in the HTML. This leads to very low Time to Visible metrics, but...

Once the client receives the HTML, a process called hydration starts, where the page is recreated client side, this time with event listeners and interactivity built-in. This is the Time to Interactive metric.

When does a page exist? AFAIK, during dev mode, the file system is matched aggressively, for example if you reach a page that doesn't exist, next.js does checks every couple of seconds to see if it has been created, and if so reload the browser to show the page. 

During build time, pages that can be optimised, are built. These includes `getStaticProps` and pages without any data fetching function. Let's ignore `getInitialProps` for arguments sake. 

Then at runtime, dynamic pages that did implement `fallback: true` or `fallback: 'blocking'`, have the ability to be generated as part of the runtime. That is to say that if you generated `post/0` and `post/1` at build time, landing them will serve them directly, but for `post/2`, you'll need to wait for the server to ship a fallback (when true) or the entire page (blocking). 

Alternatively pages with revalidate also have some mechanism to be regenerated (ISR). And you can trigger this on demand via the revalidate API.

Also, pages with `getServerSideProps`, both at dev-time and runtime,  are rendered on demand in two ways. You land on the page, then the entire HTML + JSON is shipped to the client, or, you SPA transition into the page, with `next/link`, so the JS and the JSON (result of calling `getServerSideProps` on the server) are downloaded, and are client side rendered. During build-time these pages are not executed.

> 4. The above questions are when you do it in dev mode, but I wonder if there is a difference in rendering with the above questions when using prod.

I think I've covered most of it? It is not a short, one liner, subject. I recall giving a presentation on this, and it took about 15 minutes to sketch a rough map of everything that's possible on such a powerful, yet intuitive framework. Please ask away if I was not clear in some parts, or if I've missed details.


## Edit

SPA transitions can also be triggered via `next/router` methods.

**Reply by zcarc** [reply]
thanks for the answer.
